### PR TITLE
Add a disassembler

### PR DESCRIPTION
**Pull request by zenhack**
I wrote this while debugging the multi-assets work. There are some
things left to do:

- I need to properly document its interface, mainly the exact format of
  the return value.
- There is some code to try to detect when a PUSH is pushing a value
  corespodning to the address of a JUMPDEST, and record this fact, but
  It doesn't appear to be working; I should either figure out why or
  just punt and drop this feature.

...but I wanted to get this recorded somewhere before calling it a day.

### CODE REVIEW

**Comment by fare** [comment]
Looks generally fine to me. See @AlexKnauth 's suggestion for a better handling of PUSH. String manipulation is better avoided.

**Comment by zenhack** [comment]
Alright, I've:

- Added a test
- Incorporated your suggestion, pulling in code from @AlexKnauth implementation on discord
- Dropped the WIP on detecting label references; we can add that later if we want.
- Added documentation.

Marking this as not-a-draft.

File: assembly.ss [hunk 1]
```diff
@@ -100,7 +102,47 @@
...
+         (name (vector-ref rev-opcodes opcode))
+         (push-amt (push-code-amount opcode)))
+        (cond
+          ((not (symbol? name))
+           (cons ['invalid opcode] (loop (cdr data))))
+          (push-amt
+           (let ()
+             (def rest (cdr data))
+             (def arg-bytes (take rest push-amt))
+             (def arg-hex (string-append "#x" (hex-encode (list->u8vector arg-bytes))))
```
Comment (fare) [inline assembly.ss:129]: Since we're in Ethereum land, I'd use "0x" instead of "#x" as the prefix.

File: assembly.ss [hunk 2]
```diff
@@ -100,7 +102,47 @@
...
+;; of instructions, of the form:
+;;
+;; - A symbol for the instruction mnemonic, for most instructions.
+;; - (PUSH* number string) for PUSH* instructions (where * is the size of the argument).
+;;   The (first) numeric argument is the actual value that is pushed. The second argument
+;;   is the same, but encoded as a hexideicmal scheme literal (e.g. "#x0ab1"); having
+;;   both forms readily available is sometimes helpful when debugging.
+(def (disassemble bytes)
+  (def labels (make-hash-table))
+  (let loop ((data (u8vector->list bytes)))
```
Comment (fare) [inline assembly.ss:115]: I would be more efficient to loop with an index.

Considering that a program is 24KB max, that performance isn't necessary to squeeze right now — however, that deserves at least a TODO comment in the code.

**Review by fare — APPROVED** [review]
LGTM modulo minor nits

**Comment by fare** [comment]
Oh, and the tests should probably involve such degenerate cases as pushing extra zeroes in front, as in `PUSH3 42 "0x00002a"`

**Comment by zenhack** [comment]
Done.

**Pull request merged**